### PR TITLE
Security: Support cert reload

### PR DESCRIPTION
**Pull request by hunterlxt**
* Support for using user-supplied fetcher to dynamically provide certificates for servers. The fetcher will be called during initialization and addition for handshakers.

* Move certificates from integration tests code into files to reduce compilation time.

### CODE REVIEW

**Comment by hunterlxt** [comment]
@BusyJay The created `CertUsrData` is not destroyed. When the server is shut down, this may cause a slight memory leak. Do you think it's better put its life management into `Server`?

File: src/security/credentials.rs [hunk 1]
```diff
@@ -65,6 +67,20 @@ pub enum CertificateRequestType {
         GRPC_SSL_REQUEST_AND_REQUIRE_CLIENT_CERTIFICATE_AND_VERIFY as u32,
 }
 
+/// User-provided callback function for reload cert. [`None`] indicates that
+/// no reloading is needed, and [`Some(ServerCredentialsBuilder)`] indicates
+/// that reloading is needed.
+pub trait UsrFetcher {
```
Comment (overvenus) [inline src/security/credentials.rs]:
```suggestion
pub trait UserFetcher {
```

Typo?

File: tests-and-examples/tests/cases/credential.rs [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+        let mut req = HelloRequest::default();
+        req.set_name("world".to_owned());
+        let reply = client.say_hello(&req).expect("rpc");
+        assert_eq!(reply.get_message(), "Hello world");
+    }
+}
+
+#[allow(dead_code)]
+static CA_CRT: &str = "-----BEGIN CERTIFICATE-----
+MIIDRTCCAi2gAwIBAgIUGrzcBxX5bVapB86XYGPetE10wiEwDQYJKoZIhvcNAQEL
```
Comment (overvenus) [inline tests-and-examples/tests/cases/credential.rs]: Could you move them into files? Large string literal increases compile time.

File: src/security/credentials.rs (hunk 1 above)
Comment (hunterlxt) [reply src/security/credentials.rs]: Fix

File: tests-and-examples/tests/cases/credential.rs (hunk 2 above)
Comment (hunterlxt) [reply tests-and-examples/tests/cases/credential.rs]: Fix

File: src/security/credentials.rs [hunk 3]
```diff
@@ -125,6 +190,57 @@ impl ServerCredentialsBuilder {
         self
     }
 
+    /// Finalize the [`ServerCredentialsBuilder`] with a user-defined fetcher
+    /// and build the [`ServerCredentials`].
+    pub fn build_with_fetcher(
```
Comment (BusyJay) [inline src/security/credentials.rs]: It's unnecessary to build fetcher using a builder. You can move it to server::Binder.

File: src/security/credentials.rs [hunk 4]
```diff
@@ -65,6 +68,22 @@ pub enum CertificateRequestType {
...
+/// no reloading is needed, and [`Some(ServerCredentialsBuilder)`] indicates
+/// that reloading is needed.
+pub trait UserFetcher {
+    fn fetch(&mut self)
+        -> std::result::Result<Option<ServerCredentialsBuilder>, Box<dyn StdError>>;
+}
+
+#[repr(C)]
+pub struct CertUserData {
+    initial_cert_cfg: Option<Box<grpc_ssl_server_certificate_config>>,
```
Comment (BusyJay) [inline src/security/credentials.rs]: It's unnecessary to store initial certificates. `UserFetcher` should be able to handle it properly.

File: src/security/credentials.rs [hunk 5]
```diff
@@ -65,6 +68,22 @@ pub enum CertificateRequestType {
         GRPC_SSL_REQUEST_AND_REQUIRE_CLIENT_CERTIFICATE_AND_VERIFY as u32,
 }
 
+/// User-provided callback function for reload cert. [`None`] indicates that
+/// no reloading is needed, and [`Some(ServerCredentialsBuilder)`] indicates
+/// that reloading is needed.
+pub trait UserFetcher {
```
Comment (BusyJay) [inline src/security/credentials.rs]: `UserFetcher` is unclear. How about `ServerCredentialsFetcher`?

File: src/security/credentials.rs [hunk 6]
```diff
@@ -170,18 +289,27 @@ impl Drop for ServerCredentialsBuilder {
 ///
 /// Use [`ServerCredentialsBuilder`] to build a [`ServerCredentials`].
 pub struct ServerCredentials {
-    creds: *mut grpc_server_credentials,
+    c_creds: *mut grpc_server_credentials,
+    cert_user_data: Option<Box<CertUserData>>,
```
Comment (BusyJay) [inline src/security/credentials.rs]: `UserFetcher` is not a part of `ServerCredentials`, it produces credentials. It' not appropriate to put it here.

Comment (hunterlxt) [reply src/security/credentials.rs]: Already put it into `ServerBuilder` and `Server`.

File: src/security/credentials.rs (hunk 5 above)
Comment (hunterlxt) [reply src/security/credentials.rs]: Fixed

File: src/security/credentials.rs (hunk 4 above)
Comment (hunterlxt) [reply src/security/credentials.rs]: The initial cert is convenient for the user to operate correctly. If it is completely handed over to the user, user may forget to set the initial(means using previous cert) and cannot load any certificates in the first time.
Anyway, I have removed it. Tell me if your opinion changes

File: src/security/credentials.rs (hunk 3 above)
Comment (hunterlxt) [reply src/security/credentials.rs]: Fixed

File: src/security/credentials.rs [hunk 7]
```diff
@@ -80,11 +96,59 @@ fn clear_key_securely(key: &mut [u8]) {
...
+    }
+    GRPC_SSL_CERTIFICATE_CONFIG_RELOAD_NEW
+}
+
 /// [`ServerCredentials`] factory in order to configure the properties.
 pub struct ServerCredentialsBuilder {
     root: Option<CString>,
     key_cert_pairs: Vec<grpcio_sys::grpc_ssl_pem_key_cert_pair>,
     cer_request_type: CertificateRequestType,
+    force_client_auth: bool,
```
[the remark this answers is not in the document]
Comment (hunterlxt) [reply src/security/credentials.rs]: Although this field is reserved in other languages (cpp, python...), it has no effect if it is removed.

File: src/security/credentials.rs [hunk 8]
```diff
@@ -65,9 +67,17 @@ pub enum CertificateRequestType {
         GRPC_SSL_REQUEST_AND_REQUIRE_CLIENT_CERTIFICATE_AND_VERIFY as u32,
 }
 
+/// User-provided callback function for reload cert. [`None`] indicates that
+/// no reloading is needed, and [`Some(ServerCredentialsBuilder)`] indicates
+/// that reloading is needed.
+pub trait ServerCredentialsFetcher {
+    fn fetch(&mut self)
+        -> std::result::Result<Option<ServerCredentialsBuilder>, Box<dyn StdError>>;
```
[the remark this answers is not in the document]
Comment (BusyJay) [reply src/security/credentials.rs]: Why not return `Result<Option<ServerCredentials>>`?

File: src/server.rs [hunk 9]
```diff
@@ -303,14 +320,14 @@ impl ServerBuilder {
...
+            for binder in self.binders.iter_mut() {
                 let bind_port = binder.bind(server);
                 if bind_port == 0 {
                     grpc_sys::grpc_server_destroy(server);
-                    return Err(Error::BindFail(binder.host, binder.port));
+                    return Err(Error::BindFail(binder.host.clone(), binder.port));
                 }
 
-                bind_addrs.push((binder.host, bind_port as u16));
+                bind_addrs.push((binder.host.clone(), bind_port as u16));
```
Comment (BusyJay) [inline src/server.rs]: How about just using `binders`?

File: src/security/credentials.rs [hunk 10]
```diff
@@ -65,9 +67,17 @@ pub enum CertificateRequestType {
...
+}
+
 impl CertificateRequestType {
     #[inline]
-    fn to_native(self) -> grpc_ssl_client_certificate_request_type {
+    pub fn to_native(self) -> grpc_ssl_client_certificate_request_type {
```
[the remark this answers is not in the document]
Comment (hunterlxt) [reply src/security/credentials.rs]: Server.rs will use it

File: src/security/credentials.rs [hunk 11]
```diff
@@ -80,6 +90,46 @@ fn clear_key_securely(key: &mut [u8]) {
...
 
+pub unsafe extern "C" fn server_cert_fetcher_wrapper(
+    user_data: *mut std::os::raw::c_void,
+    config: *mut *mut grpc_ssl_server_certificate_config,
+) -> grpc_ssl_certificate_config_reload_status {
+    if user_data.is_null() {
+        panic!("fetcher user_data must be set up!");
+    }
+    #[allow(clippy::borrowed_box)]
+    let f: &mut Box<dyn ServerCredentialsFetcher> =
```
[the remark this answers is not in the document]
Comment (hunterlxt) [reply src/security/credentials.rs]: `&mut core::ffi::c_void` doesn't implement `std::convert::AsMut<_>`

File: src/security/credentials.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (hunterlxt) [reply src/security/credentials.rs]: `grpcio::Error` represents those errors about grpc like RPCfailure, RemoteStopped, QueueShutdown. But in user fetcher, it should return a generic error type. Mainly io error that can't read the file successfully.

File: src/server.rs [hunk 12]
```diff
@@ -331,28 +348,60 @@ impl ServerBuilder {
...
+        }
+
+        /// Bind to an address with fetchers provided by users for secure connection.
+        ///
+        /// This function can be called multiple times to bind to multiple ports.
+        pub fn bind_with_fetcher<S: Into<String>>(
+            mut self,
+            host: S,
+            port: u16,
+            user_data: Box<dyn ServerCredentialsFetcher + Send>,
```
[the remark this answers is not in the document]
Comment (hunterlxt) [reply src/server.rs]: Fixed

File: src/server.rs (hunk 9 above)
Comment (hunterlxt) [reply src/server.rs]: Fixed

**Comment by hunterlxt** [comment]
@BusyJay PTAL for comments

File: src/security/credentials.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (BusyJay) [reply src/security/credentials.rs]: A complete version: `(&mut *(user_data as *mut Box<dyn ServerCredentialsFetcher>)).as_mut()`

File: src/security/credentials.rs [hunk 13]
```diff
@@ -170,18 +222,28 @@ impl Drop for ServerCredentialsBuilder {
...
     pub fn as_mut_ptr(&mut self) -> *mut grpc_server_credentials {
-        self.creds
+        self.c_creds
     }
 }
 
 impl Drop for ServerCredentials {
     fn drop(&mut self) {
-        unsafe { grpc_sys::grpc_server_credentials_release(self.creds) }
+        if !self.c_creds.is_null() {
```
Comment (BusyJay) [inline src/security/credentials.rs]: It should be checked before constructing an instance rather than here.

File: src/server.rs [hunk 14]
```diff
@@ -518,7 +564,7 @@ impl Drop for Server {
 
 impl Debug for Server {
     fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
-        write!(f, "Server {:?}", self.core.bind_addrs)
+        write!(f, "Server {:?}", self.bind_addrs().collect::<Vec<(_, _)>>())
```
Comment (BusyJay) [inline src/server.rs]: How about just `write!(f, "Server {:?}", self.core.binders.as_slice())`?

File: tests-and-examples/tests/cases/credential.rs [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+
+impl ServerCredentialsFetcher for DataInitial {
+    fn fetch(&mut self) -> Result<Option<ServerCredentialsBuilder>, Box<dyn std::error::Error>> {
+        if self.flag {
+            return Ok(None);
+        }
+        let mut pd_crt = String::new();
+        fs::File::open("certs/pd.crt")?.read_to_string(&mut pd_crt)?;
+        let mut pd_key = String::new();
+        fs::File::open("certs/pd.key")?.read_to_string(&mut pd_key)?;
```
Comment (BusyJay) [inline tests-and-examples/tests/cases/credential.rs]: There is no "pd" in gRPC.

File: tests-and-examples/tests/cases/credential.rs [hunk 16]
```diff
@@ -0,0 +1,170 @@
...
+        let mut resp = HelloReply::default();
+        resp.set_message(msg);
+        let f = sink
+            .success(resp)
+            .map_err(move |e| panic!("failed to reply {:?}", e));
+        ctx.spawn(f)
+    }
+}
+
+fn read_root_crt() -> Result<String, std::io::Error> {
```
Comment (BusyJay) [inline tests-and-examples/tests/cases/credential.rs]: Can these methods be reused?

File: tests-and-examples/tests/cases/credential.rs [hunk 17]
```diff
@@ -0,0 +1,170 @@
...
+fn read_server2_creds() -> Result<(String, String), std::io::Error> {
+    let mut server2_crt = String::new();
+    let mut server2_key = String::new();
+    fs::File::open("certs/server2.crt")?.read_to_string(&mut server2_crt)?;
+    fs::File::open("certs/server2.key")?.read_to_string(&mut server2_key)?;
+    Ok((server2_crt, server2_key))
+}
+
+struct DataReload {
+    flag: Arc<Mutex<(bool, bool)>>,
```
Comment (BusyJay) [inline tests-and-examples/tests/cases/credential.rs]: Using `AtomicBool` will be easier.

File: tests-and-examples/tests/cases/credential.rs [hunk 18]
```diff
@@ -0,0 +1,170 @@
...
+
+struct DataMeetFail {
+    flag: Arc<Mutex<bool>>,
+}
+
+impl ServerCredentialsFetcher for DataMeetFail {
+    fn fetch(&self) -> Result<Option<ServerCredentialsBuilder>, Box<dyn std::error::Error>> {
+        let mut guard_flag = self.flag.lock().unwrap();
+        if *guard_flag {
+            // Should return io::Error here.
```
Comment (BusyJay) [inline tests-and-examples/tests/cases/credential.rs]: Any updates?

File: tests-and-examples/tests/cases/credential.rs [hunk 19]
```diff
@@ -0,0 +1,170 @@
...
+                .add_cert(server2_crt.into(), server2_key.into())
+                .root_cert(root, CertificateRequestType::DontRequestClientCertificate);
+            (*guard_flag).1 = true;
+            Ok(Some(new_cred))
+        }
+    }
+}
+
+struct DataMeetFail {
+    flag: Arc<Mutex<bool>>,
```
Comment (BusyJay) [inline tests-and-examples/tests/cases/credential.rs]: The flag seems not change.

File: tests-and-examples/tests/cases/credential.rs [hunk 20]
```diff
@@ -0,0 +1,146 @@
...
+                reload: AtomicBool::new(false),
+            }),
+            CertificateRequestType::DontRequestClientCertificate,
+        )
+        .build()
+        .unwrap();
+    server.start();
+    let port = server.bind_addrs().next().unwrap().1;
+
+    // To connect the server whose CN is "remotehost"
```
Comment (BusyJay) [inline tests-and-examples/tests/cases/credential.rs]: I think it should be tested multiple times to see the certificates are really changing without breaking existing connections.

**Comment by hunterlxt** [comment]
@overvenus PTAL

File: tests-and-examples/tests/cases/credential.rs [hunk 21]
```diff
@@ -0,0 +1,160 @@
...
+    let ch = ChannelBuilder::new(env.clone())
+        .override_ssl_target("remotehost")
+        .secure_connect(&format!("localhost:{}", port.clone()), cred);
+    let client2 = GreeterClient::new(ch);
+    let mut req = HelloRequest::default();
+    req.set_name("world".to_owned());
+    let reply = client2.say_hello(&req).expect("rpc");
+    assert_eq!(reply.get_message(), "Hello world");
+
+    // To connect the server whose CN is "localhost" again.
```
Comment (BusyJay) [inline tests-and-examples/tests/cases/credential.rs]:
```suggestion
    // Existing connection is still going to work.
```

File: tests-and-examples/tests/cases/credential.rs [hunk 22]
```diff
@@ -0,0 +1,160 @@
...
+    let mut req = HelloRequest::default();
+    req.set_name("world".to_owned());
+    let reply = client1.say_hello(&req).expect("rpc");
+    assert_eq!(reply.get_message(), "Hello world");
+}
+
+#[test]
+fn test_reload_fail() {
+    let env = Arc::new(EnvBuilder::new().build());
+    let service = create_greeter(GreeterService {});
```
Comment (BusyJay) [inline tests-and-examples/tests/cases/credential.rs]: `{}` is unnecessary.

**Comment by hunterlxt** [comment]
@wujy-cs @BusyJay PTAL

**Comment by hunterlxt** [comment]
@overvenus PTAL

**Pull request merged**